File: 345.cpp
class Solution {
public:
    bool isVowels(char c)
    {
    	if(c=='a' || c=='e'|| c=='i'|| c=='o' || c=='u' || c=='A' || c=='E'|| c=='I'|| c=='O' || c=='U' )
    		return true;
    	else
    	    return false;
    }

    string reverseVowels(string s) 
    {
        int sum=-1;
        int vowels[100] = {-1};
        int half=-1;
        char temp='';
        for(int i=0;i<s.length();i++)
        {
    		if(isVowels(s[i]))
    		{
    			++sum;
    			vowels[sum]=i;
    		}
        }
        if(sum%2 > 0)
        {
            half=(sum/2)+1;
        }
        else
        	half=sum/2;
        for(int j=0;j<half;j++)
        {
        	temp=s[vowels[j]];
        	s[vowels[j]]=s[vowels[sum-j]];
        	s[vowels[sum-j]]=temp;
        }
     return s;
     }
};

File: 345.py
class Solution:
def reverseVowels(self, s):
"""
:type s: str
:rtype: str
"""
# 相當於一個有限制的字串逆序，只對字串中的母音字母進行逆序。
# Approach #1
# vowels = ('a','e','i','o','u','A','E','I','O','U')
# ss = [i for i in s if i in vowels]
# return ''.join([ss.pop() if i in vowels else i for i in list(s)  ])
# Approach #2
import re
ss = re.findall('[aeiouAEIOU]',s)
return re.sub('[aeiouAEIOU]',lambda m: ss.pop() ,s)
# Approach #3
# import re
# return re.sub('(?i)[aeiou]', lambda m, v=re.findall('(?i)[aeiou]', s): v.pop(), s)